### PR TITLE
[TACHYON-1692] Introducing wire types

### PR DESCRIPTION
**Pull request by jsimsa**
Addresses: https://tachyon.atlassian.net/browse/TACHYON-1692

The goal of this PR is to make sure that Thrift Java types are only used in service handlers. Throughout the rest of the code base, we use are own equivalents, called "wire" types. The motivation for doing this is to a protocol-independent format for that the data could be sent over the wire through, which is a pre-requisite for (easily) supporting different APIs (e.g. REST) or migrating from Thrift to gRPC. 

@calvinjia @apc999

### CODE REVIEW

**Comment by jsimsa** [comment]
addressed @yupeng9 comments on the REST API PR here

File: core/common/src/main/java/tachyon/wire/BlockInfo.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(mBlockId, mLength, mLocations);
+  }
+
+  @Override
+  public String toString() {
+    return Objects.toStringHelper(this).add("id", mBlockId).add("length", mLength)
+        .add("locations", mLocations).toString();
```
Comment (calvinjia) [inline core/common/src/main/java/tachyon/wire/BlockInfo.java]: Should we change the style of our other `toString` methods to this?

Comment (yupeng9) [inline core/common/src/main/java/tachyon/wire/BlockInfo.java]: I like the style in this PR. And I think it could be a separate PR to change the style of other `toString` methods.

File: core/common/src/test/java/tachyon/wire/BlockInfoTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.List;
+import java.util.Random;
+
+import org.codehaus.jackson.map.ObjectMapper;
+import org.junit.Assert;
+import org.junit.Test;
+
+import com.google.common.collect.Lists;
+
+public class BlockInfoTest {
```
Comment (yupeng9) [inline core/common/src/test/java/tachyon/wire/BlockInfoTest.java]: We could automate this using reflection, and add a `createExample` method to each wire class for constructing new instances. But I'm also fine with enumerating the structures in each class.

File: core/common/src/main/java/tachyon/wire/WorkerInfo.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+   */
+  public WorkerInfo setStartTimeMs(long startTimeMs) {
+    mStartTimeMs = startTimeMs;
+    return this;
+  }
+
+  /**
+   * @return thrift representation of the worker descriptor
+   */
+  public tachyon.thrift.WorkerInfo toThrift() {
```
Comment (yupeng9) [inline core/common/src/main/java/tachyon/wire/WorkerInfo.java]: What do you think of implementing an interface with a `toThrift` method for all the wire classes?

File: core/common/src/main/java/tachyon/wire/BlockInfo.java (hunk 1 above)
Comment (jsimsa) [inline core/common/src/main/java/tachyon/wire/BlockInfo.java]: Agree with @yupeng9

**Comment by jsimsa** [comment]
Jenkins, test this please.

File: core/common/src/test/java/tachyon/wire/BlockInfoTest.java (hunk 2 above)
Comment (jsimsa) [inline core/common/src/test/java/tachyon/wire/BlockInfoTest.java]: OOC, how would you test all the setters and getters using reflection? Could you perhaps give me a simple example?

File: core/common/src/main/java/tachyon/wire/WorkerInfo.java (hunk 3 above)
Comment (jsimsa) [inline core/common/src/main/java/tachyon/wire/WorkerInfo.java]: done

File: core/common/src/test/java/tachyon/wire/BlockInfoTest.java (hunk 2 above)
Comment (yupeng9) [inline core/common/src/test/java/tachyon/wire/BlockInfoTest.java]: By default, JSON serialization uses auto-detection based on the getter/setters, which means the mapper will call the getter/setters. http://www.cowtowncoder.com/blog/archives/2011/02/entry_443.html

 However, since you're using @JsonProperty here to force explicit annotations on fields, perhaps  you can change the visibility for auto-detection during tests, as in http://stackoverflow.com/questions/7105745/how-to-specify-jackson-to-only-use-fields-preferably-globally.

Lastly, since we already have all the getter/setters, @JsonProperty seems not necessary any more.

Comment (jsimsa) [inline core/common/src/test/java/tachyon/wire/BlockInfoTest.java]: I see. Removed the `@JsonProperty` annotation.

If I understand you correctly, the alternative base on reflection would consist of a single test class with a list of the wire type classes. The test class would iterate over all the test classes running the JSON and Thrift tests. Each wire type class would need to provide the createExample and also checkEquality methods.

I think I prefer having one test class per wire type. It requires a bit more boilerplate, but it avoids having to put createExample and checkEquality into the production code.

Comment (yupeng9) [inline core/common/src/test/java/tachyon/wire/BlockInfoTest.java]: A single test class, but without wire type classes. The reflection can find all classes under a package. I don't see why we need `checkEquality`, wouldn't using `.equals()` be enough? Yes, only createExample method is needed, but I think it's fine. In fact, if we ever want to generate docs for REST API in the future, then that method can be used for that too.

But as I said, I'm also fine with boilerplate for the time being, given we dont have too many structs anyway.

Comment (jsimsa) [inline core/common/src/test/java/tachyon/wire/BlockInfoTest.java]: Thanks. I will leave it as is right now.

**Comment by jsimsa** [comment]
@apc999 @yupeng9 thank you for your comments, PTAL

**Comment by calvinjia** [comment]
@jsimsa I think there are some lingering imports which are now unused.

File: core/client-internal/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 4]
```diff
@@ -270,7 +270,7 @@ public long getDefaultBlockSize() {
         ArrayList<String> hosts = new ArrayList<String>();
         List<WorkerNetAddress> addrs = Lists.newArrayList();
         // add the existing in-memory block locations first
-        for (tachyon.thrift.BlockLocation location : info.getBlockInfo().getLocations()) {
+        for (tachyon.wire.BlockLocation location : info.getBlockInfo().getLocations()) {
```
Comment (calvinjia) [inline core/client-internal/src/main/java/tachyon/hadoop/AbstractTFS.java]: Should this be imported?

File: core/common/pom.xml [hunk 5]
```diff
@@ -69,6 +69,16 @@
...
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-annotations</artifactId>
+      <version>2.5.4</version>
+    </dependency>
+    <dependency>
+      <groupId>org.codehaus.jackson</groupId>
+      <artifactId>jackson-mapper-asl</artifactId>
+      <version>1.9.13</version>
+    </dependency>
+    <dependency>
```
Comment (calvinjia) [inline core/common/pom.xml]: Are these still necessary?

File: core/common/src/main/java/tachyon/wire/BlockInfo.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+@NotThreadSafe
+public final class BlockInfo implements WireType<tachyon.thrift.BlockInfo> {
+  private long mBlockId;
+  private long mLength;
+  private List<BlockLocation> mLocations = Lists.newArrayList();
+
+  /**
+   * Creates a new instance of {@link BlockInfo}.
+   */
+  public BlockInfo() {}
```
Comment (calvinjia) [inline core/common/src/main/java/tachyon/wire/BlockInfo.java]: Is this for testing only?

**Comment by calvinjia** [comment]
@jsimsa Looks good overall, two nontrivial comments, can we remove the jackson dependency in common and is it possible to keep thrift classes from the public api (currently available through `toThrift` methods)?

File: core/client-internal/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 4 above)
Comment (jsimsa) [inline core/client-internal/src/main/java/tachyon/hadoop/AbstractTFS.java]: this is needed for disambiguation from org.apache.hadoop.fs.BlockLocation

File: core/common/pom.xml (hunk 5 above)
Comment (jsimsa) [inline core/common/pom.xml]: the annotations are not, the mapper is

File: core/common/src/main/java/tachyon/wire/BlockInfo.java (hunk 6 above)
Comment (jsimsa) [inline core/common/src/main/java/tachyon/wire/BlockInfo.java]: No. It is for instance using in BlockMaster 

`BlockInfo().setBlockId(masterBlockInfo.getBlockId()).setLength(masterBlockInfo.getLength()).setLocations(locations);`

I find this approach easier to maintain than the one with a constructor with a bunch of arguments.

**Comment by yupeng9** [comment]
LGTM

File: core/common/pom.xml (hunk 5 above)
Comment (calvinjia) [inline core/common/pom.xml]: Is it only needed in test scope?

File: core/common/src/main/java/tachyon/wire/BlockInfo.java (hunk 6 above)
Comment (calvinjia) [inline core/common/src/main/java/tachyon/wire/BlockInfo.java]: Got it. Ideally I think we should only be able to derive this object from the underlying object, but I think the code was not written in such a way before.

File: core/client-internal/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 4 above)
Comment (calvinjia) [inline core/client-internal/src/main/java/tachyon/hadoop/AbstractTFS.java]: Got it, thanks.

**Comment by jsimsa** [comment]
@calvinjia @yupeng9 to avoid leaking the Thrift class through the PublicApi, I move the toThrift methods and constructors to a Utils class ... PTAL

File: core/common/pom.xml (hunk 5 above)
Comment (jsimsa) [inline core/common/pom.xml]: done

**Comment by calvinjia** [comment]
Jenkins, test this please.

**Comment by calvinjia** [comment]
LGTM thanks.

**Comment by apc999** [comment]
LGTM thanks.
